File: api/src/main/java/com/wbrawner/budgetserver/transaction/Transaction.java
package com.wbrawner.budgetserver.transaction;

import com.wbrawner.budgetserver.budget.Budget;
import com.wbrawner.budgetserver.category.Category;
import com.wbrawner.budgetserver.recurrence.RecurringTransaction;
import com.wbrawner.budgetserver.user.User;

import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import java.time.Instant;

import static com.wbrawner.budgetserver.Utils.randomId;

@Entity
public class Transaction implements Comparable<Transaction> {
    @Id
    private final String id = randomId();
    @ManyToOne
    @JoinColumn(nullable = false)
    private final User createdBy;
    private String title;
    private String description;
    private Instant date;
    private Long amount;
    @ManyToOne
    private Category category;
    private Boolean expense;
    @ManyToOne
    @JoinColumn(nullable = false)
    private Budget budget;
    @ManyToOne
    private RecurringTransaction recurrence;

    public Transaction() {
        this(null, null, null, null, null, null, null, null, null);
    }

    public Transaction(
            String title,
            String description,
            Instant date,
            Long amount,
            Category category,
            Boolean expense,
            User createdBy,
            Budget budget,
            RecurringTransaction recurrence
    ) {
        this.title = title;
        this.description = description;
        this.date = date;
        this.amount = amount;
        this.category = category;
        this.expense = expense;
        this.createdBy = createdBy;
        this.budget = budget;
        this.recurrence = recurrence;
    }

    public String getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Instant getDate() {
        return date;
    }

    public void setDate(Instant date) {
        this.date = date;
    }

    public Long getAmount() {
        return amount;
    }

    public void setAmount(Long amount) {
        this.amount = amount;
    }

    public Category getCategory() {
        return category;
    }

    public void setCategory(Category category) {
        this.category = category;
    }

    public Boolean getExpense() {
        return expense;
    }

    public void setExpense(Boolean expense) {
        this.expense = expense;
    }

    public User getCreatedBy() {
        return createdBy;
    }

    public Budget getBudget() {
        return budget;
    }

    public void setBudget(Budget budget) {
        this.budget = budget;
    }

    public RecurringTransaction getRecurrence() {
        return this.recurrence;
    }

    public void setRecurrence(RecurringTransaction recurrence) {
        this.recurrence = recurrence;
    }

    @Override
    public int compareTo(Transaction other) {
        return this.date.compareTo(other.date);
    }
}

File: api/src/main/kotlin/com/wbrawner/budgetserver/transaction/Transaction.kt
package com.wbrawner.budgetserver.transaction

import com.wbrawner.budgetserver.budget.Budget
import com.wbrawner.budgetserver.category.Category
import com.wbrawner.budgetserver.randomString
import com.wbrawner.budgetserver.user.User
import java.time.Instant
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne

@Entity
data class Transaction(
    @Id
    val id: String = randomString(),
    var title: String? = null,
    var description: String? = null,
    var date: Instant? = null,
    var amount: Long? = null,
    @field:ManyToOne var category: Category? = null,
    var expense: Boolean? = null,
    @field:JoinColumn(nullable = false) @field:ManyToOne val createdBy: User? = null,
    @field:JoinColumn(nullable = false) @field:ManyToOne var budget: Budget? = null
)

data class NewTransactionRequest(
    val title: String,
    val description: String? = null,
    val date: String,
    val amount: Long,
    val categoryId: String? = null,
    val expense: Boolean,
    val budgetId: String
)

data class UpdateTransactionRequest(
    val title: String? = null,
    val description: String? = null,
    val date: String? = null,
    val amount: Long? = null,
    val categoryId: String? = null,
    val expense: Boolean? = null,
    val budgetId: String? = null,
    val createdBy: String? = null
)

data class TransactionResponse(
    val id: String,
    val title: String?,
    val description: String?,
    val date: String,
    val amount: Long?,
    val expense: Boolean?,
    val budgetId: String,
    val categoryId: String?,
    val createdBy: String
) {
    constructor(transaction: Transaction) : this(
        transaction.id,
        transaction.title,
        transaction.description,
        transaction.date.toString(),
        transaction.amount,
        transaction.expense,
        transaction.budget!!.id,
        transaction.category?.id,
        transaction.createdBy!!.id
    )
}